substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/ObjectFile.java

```diff
@@ -1725,4 +1741,49 @@ public final SymbolTable getOrCreateSymbolTable() {
             return createSymbolTable();
         }
     }
+
+    /**
+     * Temporary storage for a debug context installed in a nested scope under a call.
+     * to {@link #withDebugContext}
+     */
+    private DebugContext debugContext = null;
+
+    /**
+     * Allows a function to be executed with a specific debug context in a named subscope bound to
+     * the object file and accessible to code invoked during the lifetime of the function. Invoked
+     * code may obtain access to the debug context using method {@link #debugContext}.
+     * @param context a context to be bound toin the object file for the duration of the function
+     *        execution.
+     * @param scopeName a name to be used to define a subscope current while the function is being
+     *        executed.
+     * @param t a value to be injected into the function that performs the action.
+     * @param function a function to be executed while the context is bound to the object file.
+     */
+    @SuppressWarnings("try")
+    public <T, R> R withDebugContext(DebugContext context,  String scopeName, T t, Function<T, R> function) {
```

Comment:
If the return value is not used anywhere use `java.util.function.Consumer` instead `Function` of to avoid pointless code such as
```
    // doesn't actually matter what we return
    return true;
```

Reply:
Actually you can even make that a `Runnable`. The passing of the channel like this is pointless ceremony.
Then in `com/oracle/svm/hosted/image/NativeBootImage.java` you just have
```
try (FileChannel channel = FileChannel.open(outputFile, StandardOpenOption.WRITE, StandardOpenOption.READ, StandardOpenOption.TRUNCATE_EXISTING, StandardOpenOption.CREATE)) {
    objectFile.withDebugContext(context, "ObjectFile.write", () -> {
        objectFile.write(channel);
    });
}
```
Remember this is not C++ ;-)

Reply:
Well, I did start with a Consumer and then changed to a Function just in case you might want a more future-proof API. I was thinking it might later be decided that some other step of the ObjectFile write might need tracing and we could employ this Function flavour of ``withDebugContext`` where the operation needs to return a result. I'll happily switch to Consumer/Runnable. We can always switch back to Function if the need arises. 

And don't worry I have as little urge to use C++ features as the next OpenJDK programmer (well, modulo a few project 'enthusiasts' whose love of obscure C++ truly baffles me). Now *Lisp* would be a different matter . . . ;-)